notifier/notify_windows.go

```diff
@@ -2,351 +2,342 @@ package main
+	// Failure or user dismissal of notification
+	if reason == 0 {
+		notifsByIDLock.Lock()
+		delete(notifsByID, id)
+		notifsByIDLock.Unlock()
+	}
+	return 0
+}
 
-	err = cmd.Run()
-	exitCode = cmd.ProcessState.ExitCode()
+func initialize() bool {
+	// load the c library
+	err := loadDLL()
+	if err != nil {
+		log.Errorf("notify: notification library not properly loaded %s", err)
+		return false
+	}
 
-	return
-}
+	lib.Lock()
+	defer lib.Unlock()
+
+	// Initialize all necessary string for the notification meta data
+	appNameUTF, _ := windows.UTF16PtrFromString(appName)
+	companyUTF, _ := windows.UTF16PtrFromString(company)
+	productNameUTF, _ := windows.UTF16PtrFromString(productName)
+	subProductUTF, _ := windows.UTF16PtrFromString(subProduct)
+	versionInfoUTF, _ := windows.UTF16PtrFromString(info.Version())
+
+	// they are needed as unsafe pointers
+	appNameP := unsafe.Pointer(appNameUTF)
+	companyP := unsafe.Pointer(companyUTF)
+	productNameP := unsafe.Pointer(productNameUTF)
+	subProductP := unsafe.Pointer(subProductUTF)
+	versionInfoP := unsafe.Pointer(versionInfoUTF)
+
+	// Initialize notifications
+	rc, _, err := lib.initialize.Call(uintptr(appNameP), uintptr(companyP), uintptr(productNameP), uintptr(subProductP), uintptr(versionInfoP))
+	if rc != 1 {
+		log.Errorf("notify: failed to initialize notification library %q", err)
+		return false
+	}
 
-func readWideStringAsUnicode(conn net.Conn) string {
-	defer conn.Close()
+	// Initialize action callbacks
+	callback := windows.NewCallback(portmasterNotificationActivatedCallback)
+	rc, _, err = lib.setActivatedCallback.Call(callback)
 
-	var bufferslice []byte
+	if rc != 1 {
+		log.Errorf("notify: failed to initialize activated callback %q", err)
+		return false
+	}
 
-readloop:
-	for {
-		buffer := make([]byte, 512)
+	callback = windows.NewCallback(portmasterNotificationDismissedCallback)
+	rc, _, err = lib.setDismissedCallback.Call(callback)
 
-		n, err := conn.Read(buffer)
-		switch err {
-		case nil:
-			// do nothing
-		case io.EOF:
-			break readloop
-		default:
-			log.Warningf("failed to read from pipe: %s", err)
-			return ""
-		}
-		bufferslice = append(bufferslice, buffer[:n]...)
+	if rc != 1 {
+		log.Errorf("notify: failed to initialize dismissed callback %q", err)
+		return false
 	}
 
-	dec := unicode.UTF16(unicode.LittleEndian, unicode.UseBOM).NewDecoder()
-	out, err := dec.Bytes(bufferslice)
-	if err != nil {
-		log.Warningf("failed to convert wstr to str: %s", err)
-		return ""
+	callback = windows.NewCallback(portmasterNotificationDismissedCallback)
+	rc, _, err = lib.setFailedCallback.Call(callback)
+
+	if rc != 1 {
+		log.Errorf("notify: failed to initialize failed callback %q", err)
+		return false
 	}
 
-	return string(out)
+	return true
 }
 
-func getPath(module string) (string, error) {
-	// build path to app
-	if dataDir == "" {
-		return "", fmt.Errorf("failed to get Path for %s: dataDir is empty", module)
+func isInitialized() bool {
```

Comment:
Consider using `sync.Once` for initialization, as concurrent requests will wait for the init to be done and will only then continue. Right now, an early notification would just give up because the initialization is not finished yet.